app/src/main/java/de/westnordost/streetcomplete/quests/crossing/AddCrossing.kt

```diff
@@ -0,0 +1,186 @@
+import de.westnordost.streetcomplete.data.osm.mapdata.Way
+import de.westnordost.streetcomplete.data.osm.osmquests.OsmElementQuestType
+import de.westnordost.streetcomplete.quests.YesNoQuestAnswerFragment
+import de.westnordost.streetcomplete.util.initialBearingTo
+import de.westnordost.streetcomplete.util.normalizeDegrees
+
+class AddCrossing : OsmElementQuestType<Boolean> {
+
+    private val roadsFilter by lazy { """
+        ways with
+          highway ~ primary|primary_link|secondary|secondary_link|tertiary|tertiary_link|unclassified|residential
+          and area != yes
+          and crossing != no
+          and (access !~ private|no or (foot and foot !~ private|no))
+    """.toElementFilterExpression() }
+
+    private val footwaysFilter by lazy { """
+        ways with
+          (highway ~ footway|steps or highway ~ path|cycleway and foot ~ designated|yes)
+          and footway != sidewalk
+          and area != yes
+          and access !~ private|no
+    """.toElementFilterExpression() }
+
+    override val commitMessage = "Add whether there is a crossing"
+    override val wikiLink = "Tag:highway=crossing"
+    override val icon = R.drawable.ic_quest_pedestrian
+
+    override fun getTitle(tags: Map<String, String>) = R.string.quest_crossing_title
+
+    override fun isApplicableTo(element: Element): Boolean? =
+        if(element !is Node || element.tags.isNotEmpty()) false else null
+
+    override fun getApplicableElements(mapData: MapDataWithGeometry): Iterable<Element> {
+
+        val roadsByNodeId = mapData.ways.asSequence()
+            .filter { roadsFilter.matches(it) }
+            .groupByNodeIds()
+        /* filter out nodes of roads that are the end of a road network (dead end), f.e.
+         * https://www.openstreetmap.org/node/280046349 or
+         * https://www.openstreetmap.org/node/56606744 */
+        roadsByNodeId.removeEndNodes()
+
+        /* require all roads at a shared node to either have no sidewalk tagging or all of them to
+         * have sidewalk tagging: If the sidewalk tagging changes at that point, it may be an
+         * indicator that this is the transition point between separate sidewalk mapping and
+         * sidewalk mapping on road-way. F.e.:
+         * https://www.openstreetmap.org/node/1839120490 */
+        val anySidewalk = setOf("both","left","right")
+        roadsByNodeId.values.removeAll { ways ->
+            if (ways.any { it.tags["sidewalk"] in anySidewalk }) {
+                !ways.all { it.tags["sidewalk"] in anySidewalk }
+            } else {
+                false
+            }
+        }
+
+        val footwaysByNodeId = mapData.ways.asSequence()
+            .filter { footwaysFilter.matches(it) }
+            .groupByNodeIds()
+        /* filter out nodes of footways that are the end of a footway, f.e.
+         * https://www.openstreetmap.org/node/1449039062 or
+         * https://www.openstreetmap.org/node/56606744 */
+        footwaysByNodeId.removeEndNodes()
```

Comment:
this will filter out also cases when footway line is split on the road, right?

Reply:
no